README.md

```diff
@@ -0,0 +1,83 @@
+This `AddOn` implementation inserts some filters in the chain in the
+proper order to ensure that ALPN concerns are handled first, and then
+the HTTP/2 protocol is handled.  The former is of interest here.
+
+## Programmatic Configuration
+
+`AddOn` has a `setup` method that is called to allow the implementation
+to do whatever one-time setup is required to make things work.  The
+`setup` override in the `Http2AddOn` takes the following actions if the
+current connection is secure (that is, it is supposed to be using ALPN).
+
+* Use the `addHandshakeListener` method of the existing
+  [`SSLBaseFilter`](https://github.com/javaee/grizzly/blob/e0e9200479851078d9cf2bad1cf29fa72f525437/modules/grizzly/src/main/java/org/glassfish/grizzly/ssl/SSLBaseFilter.java)
+  to cause an ALPN specific SSL handshake listener to be added to the
+  existing list of listeners that are invoked when an
+  [SSL Handshake](https://www.ibm.com/support/knowledgecenter/en/SSFKSJ_7.1.0/com.ibm.mq.doc/sy10660_.htm)
+  happens.
+
+    Grizzly's handshake listener has several methods, but the only one
+    used is `onStart`.  This obtains the JDK `SSLEngine` and takes
+    different action depending on the return from its
+    [`getUseClientMode()`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLEngine.html#getUseClientMode--)
+    method.  The `grizzly-npn` module provides
+    [`AlpnClientNegotiator`](https://github.com/javaee/grizzly-npn/blob/bfa03914bf4222fd22f7710d1deb352b55de0b82/api/src/main/java/org/glassfish/grizzly/npn/AlpnClientNegotiator.java)
+    and
+    [`AlpnServerNegotiator`](https://github.com/javaee/grizzly-npn/blob/bfa03914bf4222fd22f7710d1deb352b55de0b82/api/src/main/java/org/glassfish/grizzly/npn/AlpnServerNegotiator.java)
+    interfaces for these two cases.  Please see the interfaces for the
+    specification of the required actions.
+
+    If `getUseClientMode` is true,
+
+    * use the grizzly `addCloseListener` facility to install a listener
+    that calls
+    [`NegotiationSupport`](https://github.com/javaee/grizzly-npn/blob/bfa03914bf4222fd22f7710d1deb352b55de0b82/api/src/main/java/org/glassfish/grizzly/npn/NegotiationSupport.java)
+    `.removeClientNegotiator()`.  This will ensure the client negotiator
+    is removed when the connection closes.
+
+    * Call `NegotiationSupport.addNegotiator` passing the `SSLEngine`
+    and the client negotiator impl.
+
+    If `getUseClientmode` is false,
+
+    * do the same as above, but use the server negotiator impl.
+
+    By the time `onStart` is invoked, the handshake listener will have
+    been initialized with implementations of `AlpnClientNegotiator` and
+    `AlpnServerNegotiator`.  
+
+## Deployment Configuration
+
+The above programmatic configuration steps will make it so the simple
+act of including the module built from the `bootstrap` sub-module of
+this repository in the "endorsed" directory of the JVM will enable ALPN
+support in that JVM.
```

Comment:
I would recommend documenting the use of bootclasspath vs endorsed of the JVM so as to not impact all applications that are using the JVM, only those that need ALPN.

Reply:
Done.